pkg/storage/hotplug/hotplug.go

```diff
@@ -0,0 +1,205 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright The KubeVirt Authors.
+ *
+ */
+
+package hotplug
+
+import (
+	"context"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	virtv1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/client-go/kubecli"
+	"kubevirt.io/client-go/log"
+
+	"kubevirt.io/kubevirt/pkg/apimachinery/patch"
+	storagetypes "kubevirt.io/kubevirt/pkg/storage/types"
+)
+
+func HandleDeclarativeVolumes(client kubecli.KubevirtClient, vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) error {
+	if vm.Spec.UpdateVolumesStrategy != nil && *vm.Spec.UpdateVolumesStrategy == virtv1.UpdateVolumesStrategyMigration {
+		// Are there some cases we can proceed?
+		return nil
+	}
+
+	if err := patchHotplugVolumes(client, vm, vmi); err != nil {
+		log.Log.Object(vm).Errorf("failed to update hotplug volumes for vmi:%v", err)
+		return err
+	}
+
+	return nil
+}
+
+func patchHotplugVolumes(client kubecli.KubevirtClient, vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) error {
+	if vmi == nil || !vmi.IsRunning() {
+		return nil
+	}
+
+	newVmiVolumes := append(filterHotplugVMIVolumes(vm, vmi), getNewHotplugVMVolumes(vm, vmi)...)
+	newVmiDisks := append(filterHotplugVMIDisks(vm, vmi, newVmiVolumes), getNewHotplugVMDisks(vm, vmi, newVmiVolumes)...)
+
+	if equality.Semantic.DeepEqual(vmi.Spec.Volumes, newVmiVolumes) &&
+		equality.Semantic.DeepEqual(vmi.Spec.Domain.Devices.Disks, newVmiDisks) {
+		log.Log.Object(vm).V(3).Info("No hotplug volumes to patch")
+		return nil
+	}
+
+	patchSet := patch.New(
+		patch.WithTest("/spec/volumes", vmi.Spec.Volumes),
+		patch.WithTest("/spec/domain/devices/disks", vmi.Spec.Domain.Devices.Disks),
+	)
+
+	if len(vmi.Spec.Volumes) > 0 {
+		patchSet.AddOption(patch.WithReplace("/spec/volumes", newVmiVolumes))
+	} else {
+		patchSet.AddOption(patch.WithAdd("/spec/volumes", newVmiVolumes))
+	}
+
+	if len(vmi.Spec.Domain.Devices.Disks) > 0 {
+		patchSet.AddOption(patch.WithReplace("/spec/domain/devices/disks", newVmiDisks))
+	} else {
+		patchSet.AddOption(patch.WithAdd("/spec/domain/devices/disks", newVmiDisks))
+	}
+
+	patchBytes, err := patchSet.GeneratePayload()
+	if err != nil {
+		return err
+	}
+
+	_, err = client.VirtualMachineInstance(vmi.Namespace).Patch(context.Background(), vmi.Name, types.JSONPatchType, patchBytes, v1.PatchOptions{})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func filterHotplugVMIVolumes(vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) []virtv1.Volume {
```

Comment:
`filterHotplugVMIVolumes` == `getExistingHotplugVMVolumes`?

Reply:
Not exactly. This function returns the set of volumes in the VMI that have not been removed/changed in the VM definition

Reply:
correct